mlir/lib/Dialect/Rock/utility/loweringUtils.cpp

```diff
@@ -781,3 +782,66 @@ LogicalResult mlir::rock::checkLDSSize(StringAttr arch, int64_t ldsBytes) {
   const int64_t ldsSize = rock::lookupArchInfo(arch).maxSharedMemPerWG;
   return success(ldsBytes <= ldsSize);
 }
+
+// Trace an arg back to its alloc.
+static void traceGemmAllocToArgs(memref::AllocOp buffer,
+                                 const BufferDependencyAnalysis &deps,
+                                 SmallVector<BlockArgument> &args) {
+  IRRewriter rewriter(buffer.getContext());
+  std::optional<llvm::SmallVector<OpOperand *>> readersOperands =
+      deps.getReaders(buffer);
+  if (!readersOperands.has_value())
+    return;
+  for (OpOperand *readerOperand : readersOperands.value()) {
+    if (memref::CopyOp copyOp =
+            dyn_cast<memref::CopyOp>(readerOperand->getOwner())) {
+      args.push_back(cast<BlockArgument>(copyOp.getTarget()));
+      continue;
+    }
+
+    auto readOp = dyn_cast<MemoryEffectOpInterface>(readerOperand->getOwner());
+    if (!readOp)
+      continue;
+
+    SmallVector<MemoryEffects::EffectInstance> effects;
+    readOp.getEffects(effects);
+    for (const MemoryEffects::EffectInstance &effect : effects) {
+      OpOperand *writerOperand = effect.getEffectValue<OpOperand *>();
+      // Test against the write operand to guard against [MemRead, MemWrite]
+      if (writerOperand && readerOperand != writerOperand &&
+          isa<MemoryEffects::Write>(effect.getEffect())) {
+        if (memref::AllocOp writeBuffer =
+                dyn_cast<memref::AllocOp>(writerOperand->get().getDefiningOp()))
+          traceGemmAllocToArgs(writeBuffer, deps, args);
+      }
+    }
+  }
+}
+
+FailureOr<SmallVector<BlockArgument>>
+mlir::rock::traceGemmOutputToArgs(Value matC, func::FuncOp func,
+                                  OpBuilder &builder,
+                                  const BufferDependencyAnalysis &deps) {
+  if (func.getNumArguments() == 0)
+    return failure();
+
+  // trace matC to its alloc
+  FailureOr<memref::AllocOp> allocOp = findMemrefAlloc(matC);
```

Comment:
technically matC could be the func arg (whilst we might not see exercise that given how our highlevel pipelines work) but I could hand write a valid kernel as such.
So I dont think we should fail in the case?

Reply:
I see, but then I can't use BufferDependencyAnalysis because there's no buffer. In that case we can check if matC == any kernel argument directly?

Reply:
yes -- I remember making the same complaint on BufferDependencyAnalysis work. :) 
So yea checking for argument directly might be worth it prior to failing.